src/neo4j_genai/retrievers/external/weaviate/types.py

```diff
@@ -0,0 +1,77 @@
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Optional
+
+from pydantic import (
+    field_validator,
+    BaseModel,
+    PositiveInt,
+    model_validator,
+    ConfigDict,
+)
+from weaviate.client import WeaviateClient
+from weaviate.collections.classes.filters import _Filters
+
+from neo4j_genai.retrievers.utils import validate_search_query_input
+from neo4j_genai.types import Neo4jDriverModel, EmbedderModel
+
+
+class WeaviateModel(BaseModel):
+    client: WeaviateClient
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    @field_validator("client")
+    def check_client(cls, value):
+        if not isinstance(value, WeaviateClient):
+            raise TypeError(
+                "Provided client needs to be of type weaviate.client.WeaviateClient"
+            )
+        return value
+
+
+class WeaviateNeo4jRetrieverModel(BaseModel):
+    driver_model: Neo4jDriverModel
+    client_model: WeaviateModel
+    collection: str
+    id_property_external: str
+    id_property_neo4j: str
+    embedder_model: Optional[EmbedderModel]
+    return_properties: Optional[list[str]] = None
+    retrieval_query: Optional[str] = None
+
+
+class WeaviateNeo4jSearchModel(BaseModel):
+    top_k: PositiveInt = 5
+    query_vector: Optional[list[float]] = None
+    query_text: Optional[str] = None
+    weaviate_filters: Optional[_Filters] = (None,)
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    @field_validator("weaviate_filters")
+    def check_weaviate_filters(cls, value):
+        if value and not isinstance(value, _Filters):
+            raise TypeError(
+                "Provided filters need to be of type weaviate.collections.classes.filters._Filters"
+            )
+        return value
+
+    @model_validator(mode="before")
+    def check_query(cls, values):
```

Comment:
This is great! Should it be part of all retriever search model validation? (not necessarily for this PR)

Reply:
It's currently done in the vector retrievers, but not for hybrid retrievers as it's meant to be possible because it's vector + fulltext search

Reply:
Right, makes sense.